framework/src/materials/MaterialPropertyStorage.C

```diff
@@ -361,30 +366,191 @@ MaterialPropertyStorage::initProps(MaterialData & material_data,
     {
       const auto prop_id = _stateful_prop_id_to_prop_id[i];
       mat_props.setPointer(i, material_data.props(0)[prop_id].clone(n_qpoints), {});
+      mooseAssert(mat_props[i].id() == prop_id, "Inconsistent id");
     }
 }
 
 void
 dataStore(std::ostream & stream, MaterialPropertyStorage & storage, void * context)
 {
-  unsigned int max_state = storage.maxState();
-  dataStore(stream, max_state, context);
+  auto num_states = storage.numStates();
+  dataStore(stream, num_states, nullptr);
+
+  // Store the material property ID -> name map for mapping back
+  const auto & registry = storage.getMaterialPropertyRegistry();
+  std::vector<std::string> ids_to_names(registry.idsToNamesBegin(), registry.idsToNamesEnd());
+  dataStore(stream, ids_to_names, nullptr);
+
+  // Store the stateful ID -> property ID map for mapping back
+  dataStore(stream, storage._stateful_prop_id_to_prop_id, nullptr);
 
+  // Store the stateful id -> name map
+  dataStore(stream, storage._stateful_prop_names, nullptr);
+
+  // Store every property
   for (const auto state : storage.stateIndexRange())
-    dataStore(stream, storage.setProps(state), context);
+  {
+    std::size_t num_elems = storage.setProps(state).size();
+    dataStore(stream, num_elems, nullptr);
+
+    for (auto & elem_side_map_pair : storage.setProps(state))
```

Comment:
```suggestion
    for (const auto & [elem, side_map] : storage.setProps(state))
```

Reply:
Can't due to non-cost `dataStore(std::istream &, T & data, ...)`